### PR TITLE
Create a C API to create contextes and kernels.

### PR DESCRIPTION
**Pull request by ulysseB**
This PR creates a C API to build Telamon contextes and kernels. It does not yet provides methods to use the kernels.

In order to differentiate between memory and dimension ID in the generated C header, `dim::Id` is renamed to `DimId` and `mem::Id` to `MemId`.

### CODE REVIEW

File: src/ir/dimension.rs [hunk 1]
```diff
@@ -4,30 +4,31 @@ use std::fmt;
...
 /// Provides a unique identifier for iteration dimensions.
 #[derive(Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
-pub struct Id(pub u32);
+#[repr(C)]
+pub struct DimId(pub u32);
 
-impl fmt::Debug  for Id {
+impl fmt::Debug  for DimId {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "dim::Id({})", self.0)
```
Comment (Elarnon) [inline src/ir/dimension.rs]: Does it make sense to `#[derive(Debug)]` now?

File: src/ir/error.rs [hunk 2]
```diff
@@ -47,6 +56,7 @@ impl std::fmt::Display for ExpectedType {
     fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
         match self {
             ExpectedType::Integer => write!(f, "an integer type"),
+            ExpectedType::Float => write!(f, "an floating-point type"),
```
Comment (Elarnon) [inline src/ir/error.rs]: *"a floating point type"

File: src/ir/dimension.rs (hunk 1 above)
Comment (ulysseB) [reply src/ir/dimension.rs]: Indeed.

File: src/ir/error.rs (hunk 2 above)
Comment (ulysseB) [reply src/ir/error.rs]: will fix.

File: telamon-capi/src/ir.rs [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+#[no_mangle]
+pub unsafe extern "C" fn telamon_ir_operand_new_parameter(
+    parameter: *const ir::Parameter,
+) -> *mut Operand {
+    let operand = ir::Operand::Param(&*parameter);
+    Box::into_raw(Box::new(Operand(operand)))
+}
+
+/// Creates an operand that returns the current index on a dimension.
+// FIXME: repr(C) ir::DimId
```
[the remark this answers is not in the document]
Comment (ulysseB) [reply telamon-capi/src/ir.rs]: yes.

File: telamon-capi/src/ir.rs [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+    return_type: *const ir::Type,
+) -> *mut Operator {
+    let operand = Box::from_raw(operand).0;
+    let operator = ir::Operator::Cast(operand, *return_type);
+    Box::into_raw(Box::new(Operator(operator)))
+}
+
+/// Creates an operator that loads a tensor stored in memory. Takes the ownership of
+/// `base_address` and creates copies of `strided_dims`, `strides` and `loaded_type`.
+/// This function also adds thei necessary address computation code to `function`.
```
[the remark this answers is not in the document]
Comment (ulysseB) [reply telamon-capi/src/ir.rs]: fixed

File: telamon-capi/src/ir.rs [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+
+/// Frees a function allocated with `telamon_ir_function_new`.
+#[no_mangle]
+pub unsafe extern "C" fn telamon_ir_function_free(function: *mut Function) {
+    std::mem::drop(Box::from_raw(function));
+}
+
+/// Adds an instruction performing the given operator in the given dimensions to the
+/// function. Writes the unique identifier of the instruction in `inst_id`. Returns
+/// `TELAMON_STATUS_OK` except if an error occures. Takes ownership of the operator
```
[the remark this answers is not in the document]
Comment (ulysseB) [reply telamon-capi/src/ir.rs]: fixed

File: telamon-capi/src/ir.rs [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+    TELAMON_STATUS_OK
+}
+
+/// Opaque type that abstracts away the lifetime parameter of `ir::Size` so cbindgen
+/// can generate bindings.
+pub struct Size(ir::Size<'static>);
+
+/// Create a size equal to:
+/// ```
+/// const_factor * paramr_factors[0] * .. * param_factors[num_params-1] / const_divisor
```
[the remark this answers is not in the document]
Comment (ulysseB) [reply telamon-capi/src/ir.rs]: fixed

File: telamon-capi/src/ir.rs [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+    t: *const ir::Type,
+    value: libc::int64_t,
+) -> *mut Operand {
+    unwrap_or_exit!(ir::TypeError::check_integer(*t), null);
+    let type_len = unwrap!((*t).len_byte()) as u16;
+    let operand = ir::Operand::new_int(value.into(), 8*type_len);
+    Box::into_raw(Box::new(Operand(operand)))
+}
+
+/// Creates a constant floating point operand. The provided type must be an float type.
```
[the remark this answers is not in the document]
Comment (ulysseB) [reply telamon-capi/src/ir.rs]: fixed.

File: telamon-capi/src/ir.rs [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+use std::cell::RefCell;
+use telamon::ir;
+
+pub use telamon::ir::op::Rounding;
+
+thread_local! {
+    static ERROR: RefCell<Option<ir::Error>> = RefCell::new(None);
+}
+
+const TELAMON_STATUS_OK: u32 = 1;
```
[the remark this answers is not in the document]
Comment (ulysseB) [reply telamon-capi/src/ir.rs]: fixed

**Pull request merged**